kdc/kerberos5.c

```diff
@@ -154,209 +154,271 @@ _kdc_is_anon_request(const KDC_REQ *req)
-    krb5_principal request_princ;
-    krb5_error_code ret;
-    krb5_salt def_salt;
-    krb5_enctype enctype = ETYPE_NULL;
-    const krb5_enctype *p;
-    Key *key = NULL;
-    size_t i, k, m;
-
-    if (is_preauth && (flags & KFE_USE_CLIENT) &&
-        r->client->flags.synthetic)
+    const krb5_enctype *outer;
+    const krb5_enctype *inner;
+    size_t nouter, ninner;
+    int weak_allowed = 0;
+    int weak_offered = 0;
+    size_t o, i, k; /* iterators for outer, inner, keys */
+    size_t limit = 200; /* How hard we'll try */
+
+    /* Synthetic clients have no symmetric keys for PAC-ENC-* pre-auth */
+    if (is_preauth && use_client && r->client->flags.synthetic)
         return KRB5KDC_ERR_ETYPE_NOSUPP;
 
-    if ((flags & KFE_USE_CLIENT) && !r->client->flags.synthetic) {
-	princ = r->client;
-	request_princ = r->client_princ;
-    } else {
-	princ = r->server;
-	request_princ = r->server->principal;
+    if (!cached_kerberos_enctypes) {
+        /*
+         * NOTE: This is either all strong enctypes only, or all enctypes,
+         *       including weak ones, depending on configuration.
+         */
+        cached_kerberos_enctypes = krb5_kerberos_enctypes(r->context);
+        while (cached_kerberos_enctypes[ncached_kerberos_enctypes] != ETYPE_NULL)
+             ncached_kerberos_enctypes++;
     }
 
-    use_strongest_session_key =
-	is_preauth ? r->config->preauth_use_strongest_session_key
-            : (is_tgs ? r->config->tgt_use_strongest_session_key :
-		        r->config->svc_use_strongest_session_key);
-
-    /* We'll want to avoid keys with v4 salted keys in the pre-auth case... */
-    ret = krb5_get_pw_salt(r->context, request_princ, &def_salt);
-    if (ret)
-	return ret;
-
-    ret = KRB5KDC_ERR_ETYPE_NOSUPP;
+    if (use_client)
+        princ = r->client;
+    else
+	princ = r->server;
 
     /*
-     * Pick an enctype that is in the intersection of:
-     *
-     *  - permitted_enctypes (local policy)
-     *  - requested enctypes (KDC-REQ-BODY's etype list)
-     *  - the client's long-term keys' enctypes
-     *    OR
-     *    the server's configured etype list
-     *
-     * There are two sub-cases:
-     *
-     *  - use local enctype preference (local policy)
-     *  - use the client's preference list
+     * If the client is synthetic and the synthesized entry doesn't list
+     * enctypes, then use the server's entry for figuring out what is
+     * supported.
      */
+    if (use_client && princ->flags.synthetic &&
+        (!princ->etypes || !princ->etypes->val || !princ->etypes->len))
+        princ = r->server;
+
+    if (is_preauth && r->config->preauth_use_strongest_session_key)
+        use_strongest_session_key = TRUE;
+    else if (is_tgs)
+        use_strongest_session_key = r->config->tgt_use_strongest_session_key;
+    else
+        use_strongest_session_key = r->config->svc_use_strongest_session_key;
+
+    /* AFS may still be stuck using 1DES (XXX Make this configurable!) */
```

Comment:
I think this really should be configurable, but I decided to post this PR now for your review.

Reply:
Never mind.